File: VBAudioRouter/Utils/Extensions.vb
Imports Windows.UI

Namespace Utils
    Module Extensions
        ''' <summary>
        ''' https://stackoverflow.com/a/14353572/15213858
        ''' </summary>
        ''' <returns></returns>
        <Extension()>
        Public Function Map(ByVal value As Double, ByVal fromSource As Double, ByVal toSource As Double, ByVal fromTarget As Double, ByVal toTarget As Double) As Double
            Return (value - fromSource) / (toSource - fromSource) * (toTarget - fromTarget) + fromTarget
        End Function

        <Extension>
        Public Function FindNameRecursive(Of T As FrameworkElement)(ByRef ele As FrameworkElement, name As String) As T
            For child_index = 0 To VisualTreeHelper.GetChildrenCount(ele) - 1
                Dim child As FrameworkElement = DirectCast(VisualTreeHelper.GetChild(ele, child_index), FrameworkElement)

                Dim search As T = DirectCast(child.FindName(name), T)
                If search IsNot Nothing Then Return search

                Dim recursion As T = FindNameRecursive(Of T)(child, name)
                If recursion IsNot Nothing Then Return recursion
            Next
            Return Nothing
        End Function

        <Extension>
        Public Function FindNameRecursive(ByRef ele As FrameworkElement, name As String) As FrameworkElement
            Return FindNameRecursive(Of FrameworkElement)(ele, name)
        End Function

        ''' <summary>
        ''' https://stackoverflow.com/a/24120993/15213858
        ''' </summary>
        ''' <param name="this"></param>
        <Extension>
        Public Sub BringToFront(ByRef this As FrameworkElement)
            Try
                Dim parent As Panel = DirectCast(this.Parent, Panel)
                Dim currentIndex As Integer = Canvas.GetZIndex(this)
                Dim zIndex As Integer = 0
                Dim maxZ As Integer = 0
                Dim child As FrameworkElement

                For i As Integer = 0 To parent.Children.Count - 1
                    If TypeOf parent.Children(i) Is UserControl AndAlso parent.Children(i) IsNot this Then
                        child = TryCast(parent.Children(i), UserControl)
                        zIndex = Canvas.GetZIndex(child)
                        maxZ = Math.Max(maxZ, zIndex)

                        If zIndex >= currentIndex Then
                            Canvas.SetZIndex(child, zIndex - 1)
                        End If
                    End If
                Next

                Canvas.SetZIndex(this, maxZ)
            Catch : End Try
        End Sub
    End Module

    Public Class ColorTranslator
        ''' <summary>
        ''' http://joeljoseph.net/converting-hex-to-color-in-universal-windows-platform-uwp/
        ''' </summary>
        Public Shared Function FromHex(hex As String) As Color
            hex = hex.Replace("#", String.Empty)
            Dim a As Byte = 255
            Dim r As Byte = CType(Convert.ToUInt32(hex.Substring(0, 2), 16), Byte)
            Dim g As Byte = CType(Convert.ToUInt32(hex.Substring(2, 2), 16), Byte)
            Dim b As Byte = CType(Convert.ToUInt32(hex.Substring(4, 2), 16), Byte)
            Return Color.FromArgb(a, r, g, b)
        End Function
    End Class

End Namespace

File: VBAudioRouter/Utils/Extensions.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Windows.UI;

namespace VBAudioRouter.Utils;

internal static class Extensions
{
    /// <summary>
    /// https://stackoverflow.com/a/14353572/15213858
    /// </summary>
    /// <returns></returns>
    public static double Map(this double value, double fromSource, double toSource, double fromTarget, double toTarget)
    {
        return (value - fromSource) / (toSource - fromSource) * (toTarget - fromTarget) + fromTarget;
    }

    public static T? FindNameRecursive<T>(this FrameworkElement ele, string name) where T : FrameworkElement
    {
        for (var child_index = 0; child_index <= VisualTreeHelper.GetChildrenCount(ele) - 1; child_index++)
        {
            FrameworkElement child = (FrameworkElement)VisualTreeHelper.GetChild(ele, child_index);

            var search = (T?)child.FindName(name);
            if (search != null)
                return search;

            var recursion = child.FindNameRecursive<T>(name);
            if (recursion != null)
                return recursion;
        }
        return null;
    }

    public static FrameworkElement FindNameRecursive(this FrameworkElement ele, string name)
        => ele.FindNameRecursive(name);

    /// <summary>
    /// https://stackoverflow.com/a/24120993/15213858
    /// </summary>
    /// <param name="this"></param>
    public static void BringToFront(this FrameworkElement @this)
    {
        if (@this.Parent is not Panel parent)
            return;

        int currentIndex = Canvas.GetZIndex(@this);
        int maxZ = 0;
        foreach (var child in parent.Children)
        {
            if (child == @this)
                continue;

            var zIndex = Canvas.GetZIndex(child);
            maxZ = Math.Max(maxZ, zIndex);

            if (zIndex >= currentIndex)
                Canvas.SetZIndex(child, zIndex - 1);
        }

        Canvas.SetZIndex(@this, maxZ);
    }
}

internal static class ColorTranslator
{
    /// <summary>
    /// http://joeljoseph.net/converting-hex-to-color-in-universal-windows-platform-uwp/
    /// </summary>
    public static Color FromHex(string hex)
    {
        hex = hex.Replace("#", string.Empty);
        byte a = 255;
        byte r = System.Convert.ToByte(Convert.ToUInt32(hex.Substring(0, 2), 16));
        byte g = System.Convert.ToByte(Convert.ToUInt32(hex.Substring(2, 2), 16));
        byte b = System.Convert.ToByte(Convert.ToUInt32(hex.Substring(4, 2), 16));
        return Color.FromArgb(a, r, g, b);
    }
}